### PR TITLE
Certificate submitter optimization

### PR DESCRIPTION
**Pull request by VolodymyrBabaiev**
…socket request.

### CODE REVIEW

**Review by MASOUDFOROTAN — APPROVED** [review]

File: sdk/src/test/scala/com/horizen/websocket/MainchainNodeChannelImplTest.scala [hunk 1]
```diff
@@ -387,4 +391,65 @@ class MainchainNodeChannelImplTest extends JUnitSuite with MockitoSugar {
     assertTrue("Certificate is expected to be sent.", mcRefTry.isSuccess)
     assertEquals("Certificate hash is different.", certHash, BytesUtils.toHexString(mcRefTry.get.certificateId))
   }
+
+  @Test
+  def getTopQualityCertificates(): Unit = {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/test/scala/com/horizen/websocket/MainchainNodeChannelImplTest.scala:396]: Also, we need to extend `sc_multiple_certs.py` test to verify the new logic. To be discussed.

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: qa/demo.py [hunk 2]
```diff
@@ -263,12 +263,9 @@ def run_test(self):
         sc_block_ids = generate_next_blocks(sc_node, "first node", 4)
         print("\nGenerating Withdrawal Certificate...\n")
 
-        attempts = 20
-        while mc_node.getmempoolinfo()["size"] == 0 and attempts > 0:
-            if attempts % 4 == 0:
-                print("Wait for withdrawal certificate in MC memory pool...")
-            time.sleep(10)
```
Comment (i-Alex) [inline qa/demo.py:270]: let's keep sleep at least for 1 sec.

File: qa/sc_backward_transfer.py [hunk 3]
```diff
@@ -110,11 +110,10 @@ def run_test(self):
         check_mcreference_presence(we1_1_mcblock_hash, scblock_id3, sc_node)
 
         # Wait until Certificate will appear in MC node mempool
-        attempts = 20
-        while mc_node.getmempoolinfo()["size"] == 0 and attempts > 0:
+        time.sleep(10)
+        while mc_node.getmempoolinfo()["size"] == 0 and sc_node.debug_isCertGenerationActive()["result"]["state"]:
             print("Wait for certificate in mc mempool...")
             time.sleep(10)
```
Comment (i-Alex) [inline qa/sc_backward_transfer.py:116]: can reduce to 1 sec, for example. Same in other places.

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: sdk/src/test/scala/com/horizen/websocket/MainchainNodeChannelImplTest.scala [hunk 4]
```diff
@@ -387,4 +391,108 @@ class MainchainNodeChannelImplTest extends JUnitSuite with MockitoSugar {
...
+    assertEquals("Top certificates information is different.", expectedMempoolCertQuality, mcRefTry.get.mempoolCertInfo.quality.getOrElse(0))
+    assertEquals("Top certificates information is different.", expectedMempoolCertRawCert, mcRefTry.get.mempoolCertInfo.rawCertificateHex.getOrElse(""))
+    assertEquals("Top certificates information is different.", expectedMempoolCertFee, mcRefTry.get.mempoolCertInfo.fee.getOrElse(0.0), 0.000001)
+    assertEquals("Top certificates information is different.", expectedChainCertHash, mcRefTry.get.chainCertInfo.certHash.getOrElse(""))
+    assertEquals("Top certificates information is different.", expectedChainCertQuality, mcRefTry.get.chainCertInfo.quality.getOrElse(0))
+    assertEquals("Top certificates information is different.", expectedChainCertRawCert, mcRefTry.get.chainCertInfo.rawCertificateHex.getOrElse(""))
+  }
+
+  @Test
+  def getMempoolTopQualityCertificates(): Unit = {
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply sdk/src/test/scala/com/horizen/websocket/MainchainNodeChannelImplTest.scala:451]: Done.

File: sdk/src/main/scala/com/horizen/websocket/MainchainNodeChannel.scala [hunk 5]
```diff
@@ -34,6 +34,9 @@ trait MainchainNodeChannel extends MainchainNodeApi {
   // Get block headers for given set of hashes in MC node block storage (any chain)
   def getBlockHeaders(hashes: Seq[String]): Try[Seq[MainchainHeader]]
 
+  // Get information about best certificates in mempool and chain
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply sdk/src/main/scala/com/horizen/websocket/MainchainNodeChannel.scala:37]: Done.

**Review by i-Alex — APPROVED** [review]

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 6]
```diff
@@ -196,11 +200,19 @@ class CertificateSubmitter
     }
   }
 
+  private def getCertificateTopQuality(): Int = {
+    mainchainChannel.getTopQualityCertificates(BytesUtils.toHexString(BytesUtils.reverseBytes(params.sidechainId))) match {
+      case Success(topQualityCertificates)=>
+        topQualityCertificates.mempoolCertInfo.quality.getOrElse(topQualityCertificates.chainCertInfo.quality.getOrElse(0))
```
Comment (albertog78) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:206]: ignore from block certificates the one not belonging to the expected epoch and assert that the one in the mempool is of the current epoch

**Review by albertog78 — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 7]
```diff
@@ -200,52 +200,69 @@ class CertificateSubmitter
     }
   }
 
-  private def getCertificateTopQuality(): Int = {
+  private def getCertificateTopQuality(epoch: Int): Try[Long] = Try {
     mainchainChannel.getTopQualityCertificates(BytesUtils.toHexString(BytesUtils.reverseBytes(params.sidechainId))) match {
       case Success(topQualityCertificates)=>
-        topQualityCertificates.mempoolCertInfo.quality.getOrElse(topQualityCertificates.chainCertInfo.quality.getOrElse(0))
-      case Failure(_) => 0
+        if (!topQualityCertificates.mempoolCertInfo.quality.isEmpty) {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:206]: I would suggest to add the same condition as for inchain cert: ` && topQualityCertificates.mempoolCertInfo.epoch.get == epoch`

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 8]
```diff
@@ -200,52 +200,69 @@ class CertificateSubmitter
...
+          topQualityCertificates.mempoolCertInfo.quality.get
+        } else if (!topQualityCertificates.chainCertInfo.quality.isEmpty && topQualityCertificates.chainCertInfo.epoch.get == epoch) {
+          topQualityCertificates.chainCertInfo.quality.get
+        } else {
+          0
+        }
+      case Failure(ex) => {
+        ex.printStackTrace()
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:214]: use `log.info` or `log.debug`

**Review by i-Alex — CHANGES_REQUESTED** [review]
Please, do merge from dev and resolve the conflicts. + update python tests with the case when cert is present in MC mempool and then SC block is generated.

**Review by i-Alex — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**